src/analyser.cpp

```diff
@@ -0,0 +1,1632 @@
+                                               return variable->mType == AnalyserInternalVariable::Type::OVERCONSTRAINED;
+                                           })
+                                           != std::end(mInternalVariables);
+
+        if (hasUnderconstrainedVariables) {
+            if (hasOverconstrainedVariables) {
+                mModel->mPimpl->mType = AnalyserModel::Type::UNSUITABLY_CONSTRAINED;
+            } else {
+                mModel->mPimpl->mType = AnalyserModel::Type::UNDERCONSTRAINED;
+            }
+        } else if (hasOverconstrainedVariables) {
+            mModel->mPimpl->mType = AnalyserModel::Type::OVERCONSTRAINED;
+        } else if (mModel->mPimpl->mVoi != nullptr) {
+            mModel->mPimpl->mType = AnalyserModel::Type::ODE;
+        } else if (!mInternalVariables.empty()) {
+            mModel->mPimpl->mType = AnalyserModel::Type::ALGEBRAIC;
+        }
+    } else {
+        mModel->mPimpl->mType = AnalyserModel::Type::INVALID;
+    }
+
+    // Some final post-processing is now needed, if we have a valid model.
+
+    if ((mModel->mPimpl->mType == AnalyserModel::Type::ODE)
+        || (mModel->mPimpl->mType == AnalyserModel::Type::ALGEBRAIC)) {
+        // Mark some variables as external variables, if needed.
+
+        if (!externalVariables.empty()) {
+            // Check whether a variable is marked as an external variable more
+            // than once.
+
+            std::vector<VariablePtr> uniqueExternalVariables;
+            std::vector<VariablePtr> multipleExternalVariables;
+
+            for (const auto &externalVariable : externalVariables) {
+                if (std::find(uniqueExternalVariables.begin(),
+                              uniqueExternalVariables.end(),
+                              externalVariable)
+                    == uniqueExternalVariables.end()) {
+                    uniqueExternalVariables.push_back(externalVariable);
+                } else if (std::find(multipleExternalVariables.begin(),
+                                     multipleExternalVariables.end(),
+                                     externalVariable)
+                           == multipleExternalVariables.end()) {
+                    multipleExternalVariables.push_back(externalVariable);
+                }
+            }
+
+            mExternalVariables.assign(uniqueExternalVariables.begin(), uniqueExternalVariables.end());
+
+            for (const auto &multipleExternalVariable : multipleExternalVariables) {
+                auto issue = Issue::create();
+
+                issue->setDescription("Variable '" + multipleExternalVariable->name()
+                                      + "' in component '" + owningComponent(multipleExternalVariable)->name()
+                                      + "' is marked as an external variable more than once.");
+                issue->setLevel(Issue::Level::WARNING);
```

Comment:
note to self: is this really worth a warning?

Reply:
I like my code to be clean, so if I do something that is unnecessary then I want to be told about it. It's like when you declare a variable in C/C++, but never end up using it. Your compiler will warn you about it.